pkg/registry/service/rest.go

```diff
@@ -45,21 +46,104 @@ type REST struct {
 	machines    minion.Registry
 	endpoints   endpoint.Registry
 	portals     ipallocator.Interface
+	portMgr     *portAllocator
 	clusterName string
 }
 
 // NewStorage returns a new REST.
 func NewStorage(registry Registry, machines minion.Registry, endpoints endpoint.Registry, portals ipallocator.Interface,
-	clusterName string) *REST {
+	publicServicePorts *portAllocator, clusterName string) *REST {
+
 	return &REST{
 		registry:    registry,
 		machines:    machines,
 		endpoints:   endpoints,
 		portals:     portals,
+		portMgr:     publicServicePorts,
 		clusterName: clusterName,
 	}
 }
 
+// Encapsulates the semantics of a port allocation 'transaction'
+// it is better to leak ports than to double-allocate them
+// so we allocate immediately, but defer release
+// on commit we best-effort release the deferred releases
+// on rollback we best-effort release any allocations we did
+type portAllocationOperation struct {
+	pa              *portAllocator
+	allocated       []int
+	releaseDeferred []int
+	ShouldRollback  bool
+}
+
+func (op *portAllocationOperation) Init(pa *portAllocator) {
```

Comment:
The assumption was that since the port is part of spec and we're not doing a Binding yet, the user CAN request a particular port.  It's not super useful, but I'm sure someone will do something with it.  Given that, all the same arguments for sync allocation as IP allocator.